app/modules/websocket.c

```diff
@@ -0,0 +1,332 @@
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (ws->connectionState != 0 && ws->connectionState != 4) {
+    return luaL_error(L, "Websocket already connecting or connected.\n");
+  }
+  ws->connectionState = 0;
+
+  lua_pushvalue(L, 1);  // copy userdata to the top of stack to allow ref
+  data->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+
+  const char *url = luaL_checkstring(L, 2);
+  ws_connect(ws, url);
+
+  return 0;
+}
+
+static int websocketclient_send(lua_State *L) {
+  NODE_DBG("websocketclient_send is called.\n");
+
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+  luaL_argcheck(L, ws, 1, "Client websocket expected");
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    return luaL_error(L, "Client websocket is nil.\n");
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (ws->connectionState != 3) {
+    // should this be an onFailure callback instead?
+    return luaL_error(L, "Websocket isn't connected.\n");
+  }
+
+  int msgLength;
+  const char *msg = luaL_checklstring(L, 2, &msgLength);
+
+  int opCode = 1; // default: text message
+  if (lua_gettop(L) == 3) {
+    opCode = luaL_checkint(L, 3);
+  }
+
+  ws_send(ws, opCode, msg, (unsigned short) msgLength);
+  return 0;
+}
+
+static int websocketclient_close(lua_State *L) {
+  NODE_DBG("websocketclient_close.\n");
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+
+  ws_close(ws);
+  return 0;
+}
+
+static int websocketclient_gc(lua_State *L) {
+  NODE_DBG("websocketclient_gc\n");
+
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+  luaL_argcheck(L, ws, 1, "Client websocket expected");
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    return luaL_error(L, "Client websocket is nil.\n");
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (data->onConnection != LUA_NOREF) {
+    luaL_unref(L, LUA_REGISTRYINDEX, data->onConnection);
+  }
+
+  if (data->onReceive != LUA_NOREF) {
+    luaL_unref(L, LUA_REGISTRYINDEX, data->onReceive);
+  }
+
+  if (data->onClose != LUA_NOREF) {
+    if (ws->connectionState != 4) { // only call if connection open
+      lua_rawgeti(L, LUA_REGISTRYINDEX, data->onClose);
+
+      lua_pushnumber(L, -100);
+      lua_call(L, 1, 0);
+    }
+    luaL_unref(L, LUA_REGISTRYINDEX, data->onClose);
+  }
+
+  if (data->self_ref != LUA_NOREF) {
+    lua_gc(L, LUA_GCSTOP, 0); // required to avoid freeing ws_data
+    luaL_unref(L, LUA_REGISTRYINDEX, data->self_ref);
+    data->self_ref = LUA_NOREF;
+    lua_gc(L, LUA_GCRESTART, 0);
+  }
+
+  NODE_DBG("freeing lua data\n");
+  luaM_free(L, data);
+  NODE_DBG("done freeing lua data\n");
+
+  return 0;
+}
+
+static const LUA_REG_TYPE websocket_map[] =
+{
+  { LSTRKEY("createClient"), LFUNCVAL(websocket_createClient) },
+  { LNILKEY, LNILVAL }
+};
+
+static const LUA_REG_TYPE websocketclient_map[] =
+{
+  { LSTRKEY("onConnection"), LFUNCVAL(websocketclient_onConnection) },
```

Comment:
Several modules like `net` and `mqtt` only expose one single `module.on()` method where the event is identified by a string. I suggest to use [`luaL_checkoption()`](http://pgl.yoyo.org/luai/i/luaL_checkoption) for decoding instead of the if/else logic found over there.